R/derive_param_bor.R

```diff
@@ -0,0 +1,541 @@
+                                 optional = TRUE)
+
+  admiral::assert_logical_scalar(arg = missing_as_ne)
+
+  admiral::assert_function(arg = aval_fun)
+
+  admiral::assert_vars(arg = subject_keys)
+
+  admiral::assert_varval_list(arg               = set_values_to,
+                              required_elements = c("PARAMCD", "PARAM"))
+
+  admiral::assert_param_does_not_exist(dataset = dataset,
+                                       param   = rlang::quo_get_expr(set_values_to$PARAMCD))
+
+  #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+  # filter_pd and filter_source: Filter source dataset using filter_source----
+  # argument and also filter data after progressive disease with filter_pd
+  #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+  if (!is.null(source_pd)) {
+
+    # asserts on the pd data
+    source_names <- names(source_datasets)
+
+    admiral::assert_list_element(
+      list         = list(source_pd),
+      element      = "dataset_name",
+      condition    = dataset_name %in% source_names,
+      source_names = source_names,
+      message_text = paste0(
+        "The dataset names must be included in the list specified for the ",
+        "`source_datasets` parameter.\n",
+        "Following names were provided by `source_datasets`:\n",
+        admiral::enumerate(source_names, quote_fun = sQuote))
+    )
+
+    admiral::assert_s3_class(arg   = source_pd,
+                             class = "date_source")
+
+    admiral::assert_data_frame(arg = eval(rlang::parse_expr(source_pd$dataset_name)))
+
+    dataset_filter <- dataset %>%
+      filter_pd(filter          = !!enquo(filter_source),
+                source_pd       = source_pd,
+                source_datasets = source_datasets,
+                subject_keys    = subject_keys)
+
+  } else {
+
+    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    # filter_source: Filter using filter_source argument ----
+    # This would also be used to filter out records from dataset that are greater
+    # than e.g. ADSL.TRTSDT
+    # Not filtering data after progressive disease with filter_pd
+    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+    dataset_filter <- dataset %>%
+      dplyr::filter(!!rlang::enquo(filter_source))
+
+  }
+
+  # Error if filter results in 0 records
+  if (nrow(dataset_filter) == 0) {
+    err_msg <- sprintf(
+      "dataframe passed into %s argument with the filter %s has 0 records",
+      "dataset",
+      deparse(rlang::quo_get_expr(filter_source)))
+
+    rlang::abort(err_msg)
+  }
+
+  #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+  # Create Sort Order for Selection of Minimum Later ----
+  #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+  dataset_ordered <- dataset_filter %>%
+    dplyr::mutate(tmp_order = dplyr::case_when(AVALC %in% c("CR") ~ 1,
```

Comment:
Not sure this would work, for example:

SUBJID        AVALC         ADT < !!reference_date + days(ref_start_window)        ADT
001                SD                TRUE                                                                                 21 June 2022
001                SD                FALSE                                                                                28 July 2022

With this logic the first SD would be set NE and the second SD would be selected for BOR (with its associated ADT) but I assumed the first one should be selected with an ADT of 21 June 2022 (due to the fact they have a subsequent read). 

Apologies, if I am misunderstanding the requirements.

Reply:
I think selecting the second assessment is correct.

Reply:
I see, this is not how I read the issue.

_Details: checks for best overall response from PARAMCD=”OVR” (where ANL01FL=”Y”) up to/including first PD as default and sets new parameter record with AVAL/AVALC from the source record. **Set ADT as the first date where the condition is fulfilled.**_

In which case the logic you propose would work, I'll update.

Reply:
And the same is true for this @bundfussr 

```
SUBJID   AVALC             ADT            ADT < !!reference_date + days(ref_start_window)
001      SD                21 June 2022   TRUE                                                                              
001      NON-CR/NON-PD     28 July 2022   FALSE                                                                              
```

This case NON-CR/NON-PD, should be selected?

Reply:
and yes to your 2nd question about the NON-CR/NON-PD case, as the first record would be considered NE and that is the lowest in order of the IF/ELSE clauses you run through in the specs

Reply:
So this statement should read, in the issue:

_BOR is set to 'NE', in the case where the subject has **ANY** AVALC = 'SD' or 'NON-CR/NON-PD' less than xxx days after the reference date from ADSL._